iops view rights

In Cloud Director 10.5 and higher, you can disable IOPS viewing rights for tenants
iops policy is enabled and inherited from provider vdc
If you push a role to tenants without "iops view", then terraform will not be able to create a VM

I'm using terraform version:
terraform version
Terraform v1.8.4

+ provider registry.terraform.io/vmware/vcd v3.12.1


I did some more tests
my account in cloud director has "org admin" rights without "view disk iops" and "edit disk iops"

When deploying a VM from a template using terraform I get an error:
The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may contain more details.
Stack trace from the terraform-provider-vcd_v3.12.1 plugin:

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x100ebdae0]

goroutine 43 [running]:
github.com/vmware/terraform-provider-vcd/v3/vcd.updateTemplateInternalDisks(0x1400019d680, {0x1012fca20?, 0x140004f0f50?}, {0x14000728840, 0x14000190b08})
        github.com/vmware/terraform-provider-vcd/v3/vcd/resource_vcd_vapp_vm_tools.go:678 +0x320
github.com/vmware/terraform-provider-vcd/v3/vcd.createVmFromImage(0x1400019d680, {0x1012fca20, 0x140004f0f50}, {0x100ef46b9, 0x6}, {0x100effb87, 0x10})
        github.com/vmware/terraform-provider-vcd/v3/vcd/resource_vcd_vapp_vm.go:1272 +0x14b0
github.com/vmware/terraform-provider-vcd/v3/vcd.genericResourceVmCreate(0x1400019d680, {0x1012fca20, 0x140004f0f50}, {0x100ef46b9, 0x6})
        github.com/vmware/terraform-provider-vcd/v3/vcd/resource_vcd_vapp_vm.go:823 +0x1ac
github.com/vmware/terraform-provider-vcd/v3/vcd.resourceVcdStandaloneVmCreate({0x10130fca0?, 0x140004e9260?}, 0x1400019d680, {0x1012fca20, 0x140004f0f50})
        github.com/vmware/terraform-provider-vcd/v3/vcd/resource_vcd_vm.go:34 +0xc8
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).create(0x140002f1180, {0x10130fca0, 0x140004e9260}, 0x1400019d680, {0x1012fca20, 0x140004f0f50})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.31.0/helper/schema/resource.go:778 +0xe4
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0x140002f1180, {0x10130fca0, 0x140004e9260}, 0x14000726ea0, 0x1400019cd80, {0x1012fca20, 0x140004f0f50})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.31.0/helper/schema/resource.go:909 +0x884
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0x140000edbd8, {0x10130fca0?, 0x140004e90e0?}, 0x14000096050)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.31.0/helper/schema/grpc_provider.go:1074 +0xaa4
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0x140004661e0, {0x10130fca0?, 0x140004e8300?}, 0x14000554000)
        github.com/hashicorp/terraform-plugin-go@v0.20.0/tfprotov5/tf5server/server.go:859 +0x3d0
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0x1012be520, 0x140004661e0}, {0x10130fca0, 0x140004e8300}, 0x1400019c080, 0x0)
        github.com/hashicorp/terraform-plugin-go@v0.20.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:503 +0x1c0
google.golang.org/grpc.(*Server).processUnaryRPC(0x14000266000, {0x10130fca0, 0x140004e8270}, {0x101315980, 0x140003f31e0}, 0x14000420000, 0x1400046df50, 0x1019f35f8, 0x0)
        google.golang.org/grpc@v1.60.0/server.go:1372 +0xb40
google.golang.org/grpc.(*Server).handleStream(0x14000266000, {0x101315980, 0x140003f31e0}, 0x14000420000)
        google.golang.org/grpc@v1.60.0/server.go:1783 +0xc00
google.golang.org/grpc.(*Server).serveStreams.func2.1()
        google.golang.org/grpc@v1.60.0/server.go:1016 +0x5c
created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 23
        google.golang.org/grpc@v1.60.0/server.go:1027 +0x124

Error: The terraform-provider-vcd_v3.12.1 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.


My manifest:
resource "vcd_vm" "my_vm" {
  name             	= "my_vm_test_01"
  vapp_template_id 	= data.vcd_catalog_vapp_template.template.id
  memory           	= var.ram
  cpus             	= var.cpu
  org 				= var.org
  vdc 				= var.vdc


  override_template_disk {
    bus_type        = "paravirtual"
    size_in_mb      = "16384"
    bus_number      = 0
    unit_number     = 0
    storage_profile = var.vcd_storage_policy
  }
}

If you remove the override block then everything works.
With normal "org admin" where "view disk iops" and "edit disk iops" is enabled, rights the same override block works correctly

--- a/.changes/v3.14.0/1312-bug-fixes.md
+++ b/.changes/v3.14.0/1312-bug-fixes.md
@@ -1,2 +1,2 @@
-* Fix [Issue 1262](https://github.com/vmware/terraform-provider-vcd/issues/1262) Panic when creating a VM when the VCD provider
-  is configured with a user without "Organization vDC Disk: View/Edit IOPS" rights [GH-1312]
+* Fix [Issue 1262](https://github.com/vmware/terraform-provider-vcd/issues/1262): Panic when creating a VM with `vcd_vm` and `vcd_vapp_vm`
+  when the VCD provider is configured with a user without "Organization vDC Disk: View/Edit IOPS" rights [GH-1312]

--- a/.changes/v3.14.0/1312-bug-fixes.md
+++ b/.changes/v3.14.0/1312-bug-fixes.md
@@ -1,0 +1,2 @@
+* Fix [Issue 1262](https://github.com/vmware/terraform-provider-vcd/issues/1262) Panic when creating a VM when the VCD provider
+  is configured with a user without "Organization vDC Disk: View/Edit IOPS" rights [GH-1312]

--- a/vcd/resource_vcd_vapp_vm_test.go
+++ b/vcd/resource_vcd_vapp_vm_test.go
@@ -476,6 +476,7 @@ func TestAccVcdVm_WithoutIopsRights(t *testing.T) {
 	var params = StringMap{
 		"ProviderVcdOrg": providerVcdOrg1,
 		"Org":            testConfig.VCD.Org,
+		"Name":           t.Name(),
 		"Vdc":            testConfig.Nsxt.Vdc,
 		"Catalog":        testConfig.VCD.Catalog.NsxtBackedCatalogName,
 		"CatalogItem":    testConfig.VCD.Catalog.NsxtCatalogItem,

--- a/vcd/resource_vcd_vapp_vm_test.go
+++ b/vcd/resource_vcd_vapp_vm_test.go
@@ -543,6 +543,7 @@ func TestAccVcdVm_WithoutIopsRights(t *testing.T) {
 					}
 				},
 				Check: resource.ComposeTestCheckFunc(
+					// Nothing to test here, we just want to assert that no panic is thrown anymore
 					resource.TestCheckResourceAttr("vcd_vm.my_vm", "name", t.Name()),
 				),
 			},

--- a/vcd/resource_vcd_vapp_vm_test.go
+++ b/vcd/resource_vcd_vapp_vm_test.go
@@ -4,6 +4,8 @@ package vcd
 
 import (
 	"fmt"
+	"github.com/vmware/go-vcloud-director/v2/types/v56"
+	"net/url"
 	"testing"
 
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
@@ -463,6 +465,118 @@ resource "vcd_vm" "boot-image-vm" {
 
   os_type          = "sles10_64Guest"
   hardware_version = "vmx-14"
+}
+`
+
+// TestAccVcdVm_WithoutIopsRights tests VM creation with a user that does not have any View/edit IOPS rights.
+// This test will panic without fix made in <INSERT PR>.
+func TestAccVcdVm_WithoutIopsRights(t *testing.T) {
+	preTestChecks(t)
+
+	var params = StringMap{
+		"ProviderVcdOrg": providerVcdOrg1,
+		"Org":            testConfig.VCD.Org,
+		"Vdc":            testConfig.Nsxt.Vdc,
+		"Catalog":        testConfig.VCD.Catalog.NsxtBackedCatalogName,
+		"CatalogItem":    testConfig.VCD.Catalog.NsxtCatalogItem,
+		"Tags":           "vapp vm",
+	}
+	testParamsNotEmpty(t, params)
+
+	configText := templateFill(testAccCheckVcdVmWithoutIopsRights, params)
+	if vcdShortTest {
+		t.Skip(acceptanceTestsSkipped)
+		return
+	}
+
+	vcdClient := createTemporaryVCDConnection(false)
+	// Save the rights that are manipulated to restore them when the test is finished
+	var modifiedRights []types.OpenApiReference
+
+	// Restore the role with the rights that were removed, if needed, at the end of the test
+	defer func() {
+		if len(modifiedRights) > 0 {
+			role, err := vcdClient.Client.GetGlobalRoleByName("Organization Administrator")
+			if err != nil {
+				t.Fatalf("could not restore rights after %s test finished, %s", t.Name(), err)
+			}
+			err = role.AddRights(modifiedRights)
+			if err != nil {
+				t.Fatalf("could not restore rights after %s test finished, %s", t.Name(), err)
+			}
+		}
+	}()
+
+	debugPrintf("#[DEBUG] CONFIGURATION: %s\n", configText)
+	resource.Test(t, resource.TestCase{
+		ProviderFactories: buildMultipleProviders(),
+		Steps: []resource.TestStep{
+			{
+				Config: configText,
+				PreConfig: func() {
+					role, err := vcdClient.Client.GetGlobalRoleByName("Organization Administrator")
+					if err != nil {
+						if govcd.ContainsNotFound(err) {
+							t.Skipf("could not find required global role Organization Administrator")
+						}
+						t.Fatal(err)
+					}
+					filter := make(url.Values)
+					filter.Set("filter", "name==Organization vDC Disk*")
+					existingRights, err := role.GetRights(filter)
+					if err != nil {
+						t.Fatal(err)
+					}
+					if len(existingRights) == 0 {
+						return
+					}
+					for _, right := range existingRights {
+						modifiedRights = append(modifiedRights, types.OpenApiReference{
+							Name: right.Name,
+							ID:   right.ID,
+						})
+					}
+					err = role.RemoveRights(modifiedRights)
+					if err != nil {
+						t.Fatal(err)
+					}
+				},
+				Check: resource.ComposeTestCheckFunc(),
+			},
+		},
+	})
+	postTestChecks(t)
+}
+
+const testAccCheckVcdVmWithoutIopsRights = `
+data "vcd_catalog" "my-catalog" {
+  provider = {{.ProviderVcdOrg}}
+  org      = "{{.Org}}"
+  name     = "{{.Catalog}}"
+}
+
+data "vcd_catalog_vapp_template" "template" {
+  provider   = {{.ProviderVcdOrg}}
+  catalog_id = data.vcd_catalog.my-catalog.id
+  name       = "{{.CatalogItem}}"
+}
+
+resource "vcd_vm" "my_vm" {
+  provider         = {{.ProviderVcdOrg}}
+  org              = "{{.Org}}"
+  vdc              = "{{.Vdc}}"
+  name             = "{{.Name}}"
+  vapp_template_id = data.vcd_catalog_vapp_template.template.id
+  memory           = 512
+  cpus             = 1
+
+  override_template_disk {
+    bus_type        = "paravirtual"
+    size_in_mb      = "16384"
+    bus_number      = 0
+    unit_number     = 0
+    storage_profile = "*"
+  }
 }
 `
 

--- a/vcd/resource_vcd_vapp_vm_test.go
+++ b/vcd/resource_vcd_vapp_vm_test.go
@@ -510,6 +510,7 @@ func TestAccVcdVm_WithoutIopsRights(t *testing.T) {
 	debugPrintf("#[DEBUG] CONFIGURATION: %s\n", configText)
 	resource.Test(t, resource.TestCase{
 		ProviderFactories: buildMultipleProviders(),
+		CheckDestroy:      testAccCheckVcdStandaloneVmDestroy(t.Name(), params["Org"].(string), params["Vdc"].(string)),
 		Steps: []resource.TestStep{
 			{
 				Config: configText,
@@ -541,7 +542,9 @@ func TestAccVcdVm_WithoutIopsRights(t *testing.T) {
 						t.Fatal(err)
 					}
 				},
-				Check: resource.ComposeTestCheckFunc(),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("vcd_vm.my_vm", "name", t.Name()),
+				),
 			},
 		},
 	})

--- a/vcd/resource_vcd_vapp_vm_test.go
+++ b/vcd/resource_vcd_vapp_vm_test.go
@@ -469,7 +469,7 @@ resource "vcd_vm" "boot-image-vm" {
 `
 
 // TestAccVcdVm_WithoutIopsRights tests VM creation with a user that does not have any View/edit IOPS rights.
-// This test will panic without fix made in <INSERT PR>.
+// This test will panic without fix made in https://github.com/vmware/terraform-provider-vcd/pull/1312
 func TestAccVcdVm_WithoutIopsRights(t *testing.T) {
 	preTestChecks(t)
 

--- a/vcd/resource_vcd_vapp_vm_test.go
+++ b/vcd/resource_vcd_vapp_vm_test.go
@@ -472,15 +472,17 @@ resource "vcd_vm" "boot-image-vm" {
 // This test will panic without fix made in https://github.com/vmware/terraform-provider-vcd/pull/1312
 func TestAccVcdVm_WithoutIopsRights(t *testing.T) {
 	preTestChecks(t)
+	skipIfNotSysAdmin(t)
 
 	var params = StringMap{
-		"ProviderVcdOrg": providerVcdOrg1,
-		"Org":            testConfig.VCD.Org,
-		"Name":           t.Name(),
-		"Vdc":            testConfig.Nsxt.Vdc,
-		"Catalog":        testConfig.VCD.Catalog.NsxtBackedCatalogName,
-		"CatalogItem":    testConfig.VCD.Catalog.NsxtCatalogItem,
-		"Tags":           "vapp vm",
+		"ProviderVcdSystem": providerVcdSystem,
+		"ProviderVcdOrg":    providerVcdOrg1,
+		"Org":               testConfig.VCD.Org,
+		"Name":              t.Name(),
+		"Vdc":               testConfig.Nsxt.Vdc,
+		"Catalog":           testConfig.VCD.Catalog.NsxtBackedCatalogName,
+		"CatalogItem":       testConfig.VCD.Catalog.NsxtCatalogItem,
+		"Tags":              "vapp vm",
 	}
 	testParamsNotEmpty(t, params)
 
@@ -554,6 +556,14 @@ func TestAccVcdVm_WithoutIopsRights(t *testing.T) {
 }
 
 const testAccCheckVcdVmWithoutIopsRights = `
+# Acceptance tests only:
+# This data source initializes the 'testAccProvider' variable and prevents a panic
+# in 'testAccCheckVcdStandaloneVmDestroy' when running acceptance tests
+# (check TESTING.md "Tests with multiple providers" for more info)
+data "vcd_version" "dummy" {
+  provider = {{.ProviderVcdSystem}}
+}
+
 data "vcd_catalog" "my-catalog" {
   provider = {{.ProviderVcdOrg}}
   org      = "{{.Org}}"

--- a/vcd/resource_vcd_vapp_vm_test.go
+++ b/vcd/resource_vcd_vapp_vm_test.go
@@ -468,9 +468,9 @@ resource "vcd_vm" "boot-image-vm" {
 }
 `
 
-// TestAccVcdVm_WithoutIopsRights tests VM creation with a user that does not have any View/edit IOPS rights.
+// TestAccVcdVm_WithoutOrganizationVdcDiskIopsRights tests VM creation with a user that does not have any "Organization vDC Disk: View/edit" IOPS rights.
 // This test will panic without fix made in https://github.com/vmware/terraform-provider-vcd/pull/1312
-func TestAccVcdVm_WithoutIopsRights(t *testing.T) {
+func TestAccVcdVm_WithoutOrganizationVdcDiskIopsRights(t *testing.T) {
 	preTestChecks(t)
 	skipIfNotSysAdmin(t)
 
@@ -496,7 +496,8 @@ func TestAccVcdVm_WithoutIopsRights(t *testing.T) {
 	// Save the rights that are manipulated to restore them when the test is finished
 	var modifiedRights []types.OpenApiReference
 
-	// Restore the role with the rights that were removed, if needed, at the end of the test
+	// Restore the role with the rights that were removed, if needed, at the end of the test.
+	// The "modifiedRights" is set during test execution PreConfig phase
 	defer func() {
 		if len(modifiedRights) > 0 {
 			role, err := vcdClient.Client.GetGlobalRoleByName("Organization Administrator")
@@ -525,15 +526,19 @@ func TestAccVcdVm_WithoutIopsRights(t *testing.T) {
 						}
 						t.Fatal(err)
 					}
+					// Use a filter, so we only get the "Organization vDC Disk: ..." rights and skip the non-relevant ones
 					filter := make(url.Values)
 					filter.Set("filter", "name==Organization vDC Disk*")
 					existingRights, err := role.GetRights(filter)
 					if err != nil {
 						t.Fatal(err)
 					}
+					// If the rights are not present in that role, there's nothing else to do
 					if len(existingRights) == 0 {
 						return
 					}
+					// Convert the obtained rights to OpenApiReference to be able to remove them.
+					// We save them in the outer-scoped "modifiedRights" so they can be re-added later
 					for _, right := range existingRights {
 						modifiedRights = append(modifiedRights, types.OpenApiReference{
 							Name: right.Name,

--- a/vcd/resource_vcd_vapp_vm_tools.go
+++ b/vcd/resource_vcd_vapp_vm_tools.go
@@ -695,7 +695,7 @@ func updateTemplateInternalDisks(d *schema.ResourceData, meta interface{}, vm go
 		}
 
 		// Update details of internal disk for disk existing in template
-		if value, ok := internalDiskProvidedConfig["iops"]; ok {
+		if value, ok := internalDiskProvidedConfig["iops"]; ok && diskCreatedByTemplate.IopsAllocation != nil {
 			iops := int64(value.(int))
 			diskCreatedByTemplate.IopsAllocation.Reservation = iops
 		}

--- a/vcd/resource_vcd_vapp_vm_tools.go
+++ b/vcd/resource_vcd_vapp_vm_tools.go
@@ -696,6 +696,9 @@ func updateTemplateInternalDisks(d *schema.ResourceData, meta interface{}, vm go
 
 		// Update details of internal disk for disk existing in template
 		if value := internalDiskProvidedConfig["iops"]; value.(int) != 0 || diskCreatedByTemplate.IopsAllocation != nil {
+			if diskCreatedByTemplate.IopsAllocation == nil {
+				diskCreatedByTemplate.IopsAllocation = &types.IopsResource{}
+			}
 			iops := int64(value.(int))
 			diskCreatedByTemplate.IopsAllocation.Reservation = iops
 		}

--- a/vcd/resource_vcd_vapp_vm_tools.go
+++ b/vcd/resource_vcd_vapp_vm_tools.go
@@ -695,7 +695,7 @@ func updateTemplateInternalDisks(d *schema.ResourceData, meta interface{}, vm go
 		}
 
 		// Update details of internal disk for disk existing in template
-		if value, ok := internalDiskProvidedConfig["iops"]; ok && diskCreatedByTemplate.IopsAllocation != nil {
+		if value := internalDiskProvidedConfig["iops"]; value.(int) != 0 || diskCreatedByTemplate.IopsAllocation != nil {
 			iops := int64(value.(int))
 			diskCreatedByTemplate.IopsAllocation.Reservation = iops
 		}

